src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,118 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int OFFSET = 1;
+    private static final int START_INDEX = 0;
+    private static final int INITIAL_CAPACITY = 10;
+    private static final double ELEMENT_DATA_MULTIPLIER = 1.5;
+    private int listCapacity;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        listCapacity = INITIAL_CAPACITY;
+        elementData = (T[]) new Object[listCapacity];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == listCapacity) {
+            grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < START_INDEX) {
+            checkIndex(index);
+        }
+        if (size == listCapacity) {
+            grow();
+        }
+        System.arraycopy(elementData, index, elementData, index + OFFSET, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        while (size + list.size() >= listCapacity) {
+            grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            elementData[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final T oldValue = elementData[index];
+        index += 1;
+        System.arraycopy(elementData, index, elementData, index - OFFSET, size - index);
+        size--;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == null && elementData[i] == null) {
+                return removeAt(i);
+            }
+
+            if (elementData[i] != null && elementData[i].equals(element)) {
+                return removeAt(i);
+            }
```

Comment:
if you have two clauses with same logic - you need to combine conditions. 
A == B or A isn't null and a.equals(B)